### PR TITLE
  wireguard-tools: remove dependency on wireguard-go

### PR DESCRIPTION
**Pull request by ghost**
This patch removes the dependency of the wireguard package on the wireguard-go package.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by alyssais — COMMENTED** [review]
I don't understand what the point of making systemd optional is here — to avoid 26 kilobytes of unit files?

(The closure is the same either way, so it's not like it's removing a systemd dependency.)

**Comment by a-m-joseph** [comment]
I submitted this patch mainly because of the golang part.  Golang drags in a huge pile of stuff and does not build on a platform (mips64el-gnuabin32) that I am porting nixpkgs to.

The `systemdSupport` part is not a big deal for me; if you feel really strongly about this I can revert that part of the patch.  But it seemed really weird that nixpkgs has a well-established attr `systemdSupport`, yet instead of passing that through to wireguard's build process we are instead unconditionally clobbering it to `yes`.

But this is not a religious issue for me since it is an outlier case where the package has an option for systemd support that doesn't actually create a systemd dependency.  I am in the process of submitting several other patches for cases where it does affect the dependency tree.  This one is an outlier.

**Comment by alyssais** [comment]
systemdSupport is usually about removing a dependency on systemd.  I don't think the option provides any value in this case, and every option requires maintenance.

**Comment by a-m-joseph** [comment]
> systemdSupport is usually about removing a dependency on systemd. I don't think the option provides any value in this case, and every option requires maintenance.

Okay.  I have removed the systemdSupport part of the patch and updated the title.

**Comment by zx2c4** [comment]
Alternatively, could we just make this not depend on wireguard-go ever, at all? It doesn't need that.

**Comment by a-m-joseph** [comment]
> Alternatively, could we just make this not depend on wireguard-go ever, at all? It doesn't need that.

That would be great.  Let me know if I should update the patch to do that.

**Comment by zx2c4** [comment]
That would certainly be my preference.

Anyone here object to that?

**Comment by a-m-joseph** [comment]
I have updated the commit and the PR title; I guess after some amount of time passes without objections this could be merged.

**Review by zx2c4 — APPROVED** [review]
In principle this looks fine to me, though I don't know whether wireguard-go still needs to be wrapped in case it is there. I'll let somebody more nix-fluent decide this. But anyway I do like the idea of not needing wireguard-go.

**Comment by a-m-joseph** [comment]
Added a commit message; otherwise unchanged.

**Review by ius — COMMENTED** [review]
> Alternatively, could we just make this not depend on wireguard-go ever, at all? It doesn't need that.

The block is guarded by `lib.optionalString stdenv.isDarwin` and thus meant for Wireguard on OS X..

@a-m-joseph How does this affect your cross builds? It shouldn't require Go if you're cross compiling for MIPS (unless you're somehow building on Darwin and `stdenv.isDarwin` incorrectly matches `Platform`?

As it stands it looks like the only thing this will do is break WireGuard support on Darwin?

**Comment by a-m-joseph** [comment]
> As it stands it looks like the only thing this will do is break WireGuard support on Darwin?

Could you explain how it breaks wireguard on Darwin?

Also, could you explain why [this line](https://github.com/NixOS/nixpkgs/pull/160951/files#diff-72b3b0508b8d717e3d60af6320f2ce1575a3c8065b5de006f5fa115a2caf5edeL45) is in the `wireguard` expression rather than in the `wireguard-go` expression?

**Comment by a-m-joseph** [comment]
I have updated the title of this issue to clarify that it affects the `wireguard-tools` package, not the `wireguard` (i.e. kernel modules) package.  The commit itself is unchanged.

**Comment by ius** [comment]
> After some reading, it seems like Darwin lacks an in-kernel implementation of wireguard (bummer), and must use a userspace implementation instead.

Yes, this was the reasoning behind my comment. But, I was unaware of the separate `wireguard-go` package - also note the [WireGuard website](https://www.wireguard.com/install/#nix-on-darwin-userspace-go-tools) currently suggests installing `wireguard-tools` for Darwin.

> I still think the Darwin-related logic shouldn't be in `wireguard-tools`, since that is kernel/platform-independent part of wireguard.



Given there's a separate package for the Go implementation, just like you I don't understand why it's referenced here. Looks like it's introduced in https://github.com/NixOS/nixpkgs/pull/40907.

In any case, if this gets merged the WireGuard website should be updated accordingly (cc @zx2c4). I don't run Darwin myself, so maybe someone else might be able to explain (@sowelisuwi) ?)

**Comment by a-m-joseph** [comment]
> note the [WireGuard website](https://www.wireguard.com/install/#nix-on-darwin-userspace-go-tools) currently suggests installing `wireguard-tools` for Darwin.

That is recommended that for *all* nixpkgs-based users, not just Darwin.

Those instructions aren't wrong, they're incomplete.  They should read something like:

```
nix-env -iA nixpkgs.wireguard-tools nixpkgs.wireguard-go
```

> In any case, if this gets merged the WireGuard website should be updated accordingly (cc @zx2c4) to mention both packages.

I agree.

> I don't run Darwin myself

Me neither, so we should probably get somebody with a Mac to sanity-check this.

**Comment by a-m-joseph** [comment]
Ping...

**Comment by SuperSandro2000** [comment]
@alyssais @zx2c4 you still fine with this?

**Comment by hraban** [comment]
@amjoseph-nixpkgs I actually need wireguard-go in my PATH on my machine, this patch broke wg-quick for me. darwin aarch64. I didn't know of this PR and created my own, adding wireguard-go to the PATH; #251911 .

How did you test this, how certain are you that this works for wg-quick? What could I be doing wrong?

**Comment by hraban** [comment]
Hi all, any update on why this was done? Otherwise do I have your blessing to revert it, in #251911 ? wireguard-tools is currently broken on Mac unless you have wireguard-go installed in your PATH.

**Pull request merged**
